### PR TITLE
add training code + medusa_block model

### PR DESCRIPTION
**Pull request by YaelSegalFeldman**
Add training code for the two types of Medusa architectures, both with and without KL loss.

### CODE REVIEW

File: whisper_medusa/dataset/dataset.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+class ASRDataSet(torch.utils.data.Dataset):
+    """
+    ASR Dataset
+
+    Parameters
+    ----------
+    data_path : str. Path to csv file with columns: "sentence", "path".
+    """
+
```
Comment (AvivSham) [inline whisper_medusa/dataset/dataset.py:21]: update docstring to include all input args.

File: whisper_medusa/dataset/dataset.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        resampler = torchaudio.transforms.Resample(sampling_rate, resampling_to)
+        array = resampler(speech_array)[0].numpy()
+        batch["audio"] = {"array": array, "sampling_rate": resampling_to}
+        return batch
+
+    def prepare_dataset(self, batch):
+        audio = batch["audio"]
+
+        # compute log-Mel input features from input audio array
+        # todo: switch to our feature_extractor implementation which is faster.
```
Comment (AvivSham) [inline whisper_medusa/dataset/dataset.py:83]: remove todo, and consider switching to our implementation.

File: whisper_medusa/dataset/dataset.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        self.processor = processor
+
+    def _init_dataset_obj(self):
+        self.dataset_df = pd.read_csv(self.data_path)
+        self.dataset_df.sentence = self.dataset_df.sentence.fillna("")
+
+
+        # in case of debug mode, sample only 1000 samples
+        if self.debug_mode and self.split in ["val", "test"]:
+            sample_size = min(1000, len(self.dataset_df))
```
Comment (AvivSham) [inline whisper_medusa/dataset/dataset.py:57]: Let's add the number of debug examples as an argument in the parser and then pass it to the dataset object.

File: whisper_medusa/models/model.py [hunk 4]
```diff
@@ -170,11 +225,28 @@ def __init__(self, config):
...
+            self.medusa_heads = nn.ModuleList()
+            for _ in range(self.config.medusa_num_heads):
+                head_list = []
+                for _ in range(self.config.medusa_num_layers):
+                    head_list.append(
+                        MedusaResBlock(
+                            self.config.d_model, self.config.medusa_hidden_size
+                        )
+                    )
+                self.medusa_heads.append(nn.Sequential(*head_list))
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:242]: encapsulate to private func, same for base_head case and consider using dict to avoid lots of if else (now it is ok since we have only two variants).

File: whisper_medusa/models/model.py [hunk 5]
```diff
@@ -631,8 +703,12 @@ def _medusa_greedy_search(
 
                 if synced_gpus and this_peer_finished:
                     continue  # don't waste resources running the code we don't need
-                orig_logits = outputs.logits[0]
-                medusa_logits = outputs.logits[1:]
+
+                orig_logits = logits_processor(input_ids, outputs.logits[0].squeeze(0)).unsqueeze(0)
+                batch, num_medusa, seq_len, vocab_size = outputs.logits[1:].shape
+                medusa_logits = logits_processor(input_ids, outputs.logits[1:].reshape(batch* num_medusa* seq_len, vocab_size))
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:709]: cosmetic 

```suggestion
                medusa_logits = logits_processor(input_ids, outputs.logits[1:].reshape(batch * num_medusa * seq_len, vocab_size))
```

File: whisper_medusa/models/model.py [hunk 6]
```diff
@@ -1221,22 +1308,147 @@ def forward(
...
 
         if self.config.medusa_heads_type == "base_head":
             for i in range(len(self.medusa_heads)):
                 head_out = self.medusa_heads[i](hidden_states)
                 head_proj = self.whisper_model.proj_out(head_out)
+                if i == 0:
+                    base_logits = head_proj
                 medusa_logits.append(head_proj)
                 if disable_medusa:
                     break
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1323]: please add inline comment explanation

File: whisper_medusa/dataset/dataset.py (hunk 1 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/dataset/dataset.py]: done

File: whisper_medusa/dataset/dataset.py (hunk 3 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/dataset/dataset.py]: add the "debug-examples" parameters

File: whisper_medusa/models/model.py (hunk 5 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py:709]: done

File: whisper_medusa/dataset/dataset.py (hunk 2 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/dataset/dataset.py]: done

File: whisper_medusa/models/model.py [hunk 7]
```diff
@@ -1221,22 +1308,147 @@ def forward(
...
+
+                    if output_attentions:
+                        whisper_model_outputs.decoder_attentions += (
+                            medusa_block_decoder_outputs[1],
+                        )
+
+                        if whisper_model_outputs.encoder_last_hidden_state is not None:
+                            whisper_model_outputs.cross_attentions += (
+                                medusa_block_decoder_outputs[2],
+                            )
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1411]: Seems like a duplication of lines 1331-1365, create a func and call it twice.

File: whisper_medusa/models/model.py [hunk 8]
```diff
@@ -1221,22 +1308,147 @@ def forward(
...
+
 
         stack_heads_logits = torch.stack(medusa_logits, dim=0)
 
         loss = None
+        if labels is not None:
+            loss_fct = MedusaCrossEntropyLoss(
+                loss_on_original=self.config.medusa_loss_on_original
+            )
+            if hasattr(self.config, "medusa_kl_loss") and self.config.medusa_kl_loss:
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1421]: set the default value to None or false and use:
```suggestion
            if self.config.medusa_kl_loss:
```

Comment (AvivSham) [reply whisper_medusa/models/model.py:1421]: or you can use:
```suggestion
            if getattr(self.config, "medusa_kl_loss", None):
```

File: whisper_medusa/models/model.py [hunk 9]
```diff
@@ -1221,22 +1308,147 @@ def forward(
...
+                    kl_base_logits = base_logits.detach()
+
+            # move labels to correct device to enable PP
+            labels = labels.to(hidden_states.device)
+            if self.config.medusa_loss_on_original:
+                loss = loss_fct(stack_heads_logits, labels)
+                if (
+                    hasattr(self.config, "medusa_kl_loss")
+                    and self.config.medusa_kl_loss
+                ):
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1438]: same as above.

File: whisper_medusa/models/model.py [hunk 10]
```diff
@@ -1221,22 +1308,147 @@ def forward(
...
+                    kl_loss = kl_fct(stack_heads_logits,kl_base_logits)
+                    loss = loss + kl_loss
+            else:
+                loss = loss_fct(
+                    stack_heads_logits[1:], labels
+                )  # skip the first head - which is the base model
+                if (
+                    hasattr(self.config, "medusa_kl_loss")
+                    and self.config.medusa_kl_loss
+                ):
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1449]: same here.

File: whisper_medusa/models/model.py [hunk 11]
```diff
@@ -1458,3 +1670,65 @@ def generate(
 
         else:
             raise NotImplementedError("Longform generation is not supported yet")
+
+    def freeze_model_parts(self, parts_to_freeze):
+        self.whisper_model.freeze_model_parts(parts_to_freeze)
+
+    def set_output_whisper_original(self):
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1677]: make private it won't be called from outside.

File: whisper_medusa/models/model.py [hunk 12]
```diff
@@ -1458,3 +1670,65 @@ def generate(
...
+        self.whisper_model.config.output_hidden_states = True
+        self.config.output_hidden_states = True
+        self.whisper_layer = WhisperDecoderLayer(self.whisper_model.config)
+        self.whisper_layer.load_state_dict(self.whisper_model.model.decoder.layers[-1].state_dict())  # load the last layer of the whisper model
+        for param in self.whisper_layer.parameters():
+            param.requires_grad = False
+
+
+    def output_whisper_original(self, whisper_model_outputs, use_cache, past_key_values, head_mask, attention_mask, cross_attn_head_mask):
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1686]: make private.

File: whisper_medusa/models/model.py [hunk 13]
```diff
@@ -1458,3 +1670,65 @@ def generate(
...
+        with torch.no_grad():
```
Comment (AvivSham) [inline whisper_medusa/models/model.py:1687]: decorate the function it is more elegant.

File: whisper_medusa/trainer.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+    dataset_dict = get_dataset(args_i, processor)
+
+    data_collator = DataCollatorSpeechSeq2SeqWithPadding(processor)
+
+    compute_metrics_arg = (
+        partial(compute_metrics, tokenizer=processor.tokenizer)
+        if args_i.compute_wer
+        else None
+    )
```
Comment (AvivSham) [inline whisper_medusa/trainer.py:38]: lets remove this.

File: whisper_medusa/models/model.py (hunk 11 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py:1677]: but we call it from the trainer.py

File: whisper_medusa/utils/losses.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        self.kl = nn.KLDivLoss(reduction="batchmean")
+        self.loss_on_original = loss_on_original
+        self.lamda = lamda
+
+    def forward(self, logits: torch.Tensor, labels: torch.Tensor):
+        """
+        Args:
+            logits: (num_heads, batch_size, seq_len, vocab_size)
+            labels: (batch_size, seq_len, vocab_size)
+        """
```
Comment (AvivSham) [inline whisper_medusa/utils/losses.py:61]: update docstring

File: whisper_medusa/models/model.py (hunk 11 above)
Comment (AvivSham) [reply whisper_medusa/models/model.py:1677]: the freeze func was included by mistake I was referring to `set_output_whisper_original`

File: whisper_medusa/models/model.py (hunk 4 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 5 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 6 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 7 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 8 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 9 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 11 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 12 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/models/model.py (hunk 13 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/models/model.py]: done

File: whisper_medusa/trainer.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+from whisper_medusa.dataset.dataset import (
+    DataCollatorSpeechSeq2SeqWithPadding, get_dataset)
+from whisper_medusa.models import get_model
+from whisper_medusa.utils.config_and_args import get_training_args
+from whisper_medusa.utils.medusa_trainer_cls import MedusaTrainer
+from whisper_medusa.utils.metrics import compute_metrics
+from whisper_medusa.utils.utils import (parse_args, set_logger, set_seed )
+
+def count_parameters(model):
+    return sum(p.numel() for p in model.parameters() if p.requires_grad)
```
[the remark this answers is not in the document]
Comment (YaelSegalFeldman) [reply whisper_medusa/trainer.py]: done

File: whisper_medusa/trainer.py (hunk 14 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/trainer.py]: done

File: whisper_medusa/utils/__init__.py
[the remark this answers is not in the document]
Comment (YaelSegalFeldman) [reply whisper_medusa/utils/__init__.py:1]: done

File: whisper_medusa/utils/losses.py (hunk 15 above)
Comment (YaelSegalFeldman) [reply whisper_medusa/utils/losses.py:78]: done

File: whisper_medusa/utils/medusa_trainer_cls.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        model.train()
+        inputs = self._prepare_inputs(inputs)
+        
+        with self.compute_loss_context_manager():
+            loss, loss_per_head = self.compute_loss(model, inputs)
+
+        if self.args.n_gpu > 1:
+            loss = loss.mean()  # mean() to average on multi-gpu parallel training
+            # todo: make sure the mean is on the right axis
+            loss_per_head = loss_per_head.mean(0)
```
[the remark this answers is not in the document]
Comment (YaelSegalFeldman) [reply whisper_medusa/utils/medusa_trainer_cls.py]: done

File: whisper_medusa/utils/utils.py [hunk 18]
```diff
@@ -75,3 +75,273 @@ def get_device(gpu_id="0"):
...
+
+    
+    args_ = parser.parse_args()
+
+    if args_.wandb_logging:
+        if args_.resume_from_checkpoint and not args_.wandb_id is None:
+            wandb.init(
+                id=args_.wandb_id,
+                project="<YOUR_PROJECT>",
+                entity="<YOUR_ENTITY>",
```
[the remark this answers is not in the document]
Comment (YaelSegalFeldman) [reply whisper_medusa/utils/utils.py]: done

**Review by AvivNavon — APPROVED** [review]
See minor comments

**Review by AvivSham — DISMISSED** [review]
Thanks for the changes LGTM.

**Review by AvivSham — APPROVED** [review]
We can merge.

**Review by AvivSham — APPROVED** [review]
We can merge.

**Review by YaelSegalFeldman — COMMENTED** [review]
LGTM
